pkg/agent/controller/networkpolicy/fqdn.go

```diff
@@ -405,80 +409,102 @@ func (f *fqdnController) deleteRuleSelectedPods(ruleID string) error {
 
 func (f *fqdnController) onDNSResponse(
 	fqdn string,
-	responseIPs map[string]net.IP,
-	lowestTTL uint32,
+	newDNSresponseIPs map[string]ipWithTTL,
 	waitCh chan error,
 ) {
-	if len(responseIPs) == 0 {
+	if len(newDNSresponseIPs) == 0 {
 		klog.V(4).InfoS("FQDN was not resolved to any addresses, skip updating DNS cache", "fqdn", fqdn)
 		if waitCh != nil {
 			waitCh <- nil
 		}
 		return
 	}
-	// mustCacheResponse is only true if the FQDN is already tracked by this
-	// controller, or it matches at least one fqdnSelectorItem from the policy rules.
-	// addressUpdate is only true if there has been an update in IP addresses
-	// corresponded with the FQDN.
-	mustCacheResponse, addressUpdate := false, false
-	recordTTL := time.Now().Add(time.Duration(lowestTTL) * time.Second)
+
+	currentTime := time.Now()
+	ipWithTTLMap := make(map[string]ipWithTTL)
+
+	//minTimeToReQuery Establishes a maximum reference time for tracking the minimum re-query time to DNS, as IPs expire.
+	minTimeToReQuery := time.Now().Add(24 * time.Hour)
+	addressUpdate := false
+
+	addIPToCache := func(ip string, ipMeta ipWithTTL) {
+		ipWithTTLMap[ip] = ipMeta
+		minTimeToReQuery = earlierOf(ipMeta.expirationTime, minTimeToReQuery)
+	}
 
 	f.fqdnSelectorMutex.Lock()
 	defer f.fqdnSelectorMutex.Unlock()
-	oldDNSMeta, exist := f.dnsEntryCache[fqdn]
+
+	cachedDNSMeta, exist := f.dnsEntryCache[fqdn]
 	if exist {
-		mustCacheResponse = true
-		for ipStr := range responseIPs {
-			if _, ok := oldDNSMeta.responseIPs[ipStr]; !ok {
-				addressUpdate = true
-				break
+		// Data related to this domain exists in the cache.
+		// Besides presence of existing IPs, two scenarios : 1) May get New IPs 2) Some old IPs may be absent.
+
+		// check for new IPs and these new IPs need to be added with its new TTL value as received in response.
+		for ipStr, ipMeta := range newDNSresponseIPs {
+			if _, exist := cachedDNSMeta.responseIPs[ipStr]; !exist {
+				addIPToCache(ipStr, ipMeta)
 			}
 		}
-		for oldIPStr, oldIP := range oldDNSMeta.responseIPs {
-			if _, ok := responseIPs[oldIPStr]; !ok {
-				if oldDNSMeta.expirationTime.Before(time.Now()) {
-					// This IP entry has already expired and not seen in the latest DNS response.
-					// It should be removed from the cache.
+
+		for cachedIpStr, cachedIpMeta := range cachedDNSMeta.responseIPs {
+			if _, exist := newDNSresponseIPs[cachedIpStr]; !exist {
```

Comment:
```suggestion
			if newIPMeta, exist := newDNSresponseIPs[cachedIpStr]; !exist {
```
then use it at L461, instead of querying the map another time.